### PR TITLE
Implement multi insert test case for pkg agent handler - part 2

### PR DESCRIPTION
**Pull request by kevindiu**
### Description:

This PR implements pkg agent ngt multi insert test case (part 2).
I will create another PR to implement the remaining test cases.

### Environment:

- Go Version: 1.18.1
- Docker Version: 20.10.8
- Kubernetes Version: 1.22.0
- NGT Version: 1.14.3

### Types of changes:

- [ ] Bug fix [type/bug]
- [ ] New feature [type/feature]
- [x] Add tests [type/test]
- [ ] Security related changes [type/security]
- [ ] Add documents [type/documentation]
- [ ] Refactoring [type/refactoring]
- [ ] Update dependencies [type/dependency]
- [ ] Update benchmarks and performances [type/bench]
- [ ] Update CI [type/ci]

### Changes to Core Features:

- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your core changes, as applicable?
- [x] Have you successfully ran tests with your changes locally?

### Checklist:

- [x] I have read the [CONTRIBUTING](https://github.com/vdaas/vald/blob/master/CONTRIBUTING.md) document.
- [x] I have checked open [Pull Requests](https://github.com/vdaas/vald/pulls) for the similar feature or fixes?
- [x] I have added tests and benchmarks to cover my changes.
- [x] I have ensured all new and existing tests passed.
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: pkg/agent/core/ngt/handler/grpc/handler_test.go [hunk 1]
```diff
@@ -3664,6 +3664,7 @@ func Test_server_MultiInsert(t *testing.T) {
 		id        = "uuid-1"           // insert request id
 		intVecDim = 3                  // int vector dimension
 		f32VecDim = 3                  // float32 vector dimension
+		maxVecDim = 1 << 18            // reference value for testing
```
Comment (vankichi) [inline pkg/agent/core/ngt/handler/grpc/handler_test.go:3667]: could you please add a comment like "this value is temporary"?

File: pkg/agent/core/ngt/handler/grpc/handler_test.go [hunk 2]
```diff
@@ -3742,6 +3743,22 @@ func Test_server_MultiInsert(t *testing.T) {
 		return req
 	}
 
+	genSameReq := func(num int, vec []float32) *payload.Insert_MultiRequest {
```
Comment (vankichi) [inline pkg/agent/core/ngt/handler/grpc/handler_test.go:3746]:
```suggestion
	genMInsertReq := func(num int, vec []float32) *payload.Insert_MultiRequest {
```

File: pkg/agent/core/ngt/handler/grpc/handler_test.go [hunk 3]
```diff
@@ -4302,6 +4319,702 @@ func Test_server_MultiInsert(t *testing.T) {
...
+			},
+			want: want{
+				wantRes: genObjectLocations(100, name, ip),
+			},
+		},
+		{
+			name: "Boundary Value Testing case 2.2: Success to MultiInsert with min value vector (vector type is float32)",
+			args: args{
+				ctx:  ctx,
+				reqs: genSameReq(100, []float32{math.SmallestNonzeroFloat32, math.SmallestNonzeroFloat32, math.SmallestNonzeroFloat32}),
```
Comment (vankichi) [inline pkg/agent/core/ngt/handler/grpc/handler_test.go:4377]: why did you use `SmallestNonzeroFloat32`?
the value of the vector is not always positive.

File: pkg/agent/core/ngt/handler/grpc/handler_test.go [hunk 4]
```diff
@@ -4302,6 +4319,702 @@ func Test_server_MultiInsert(t *testing.T) {
...
+						&errdetails.ResourceInfo{
+							ResourceType: ngtResourceType + "/ngt.MultiInsert",
+							ResourceName: fmt.Sprintf("%s: %s(%s)", apiName, name, ip),
+						}),
+				},
+			}
+		}(),
+		func() test {
+			req := genF32Req(vector.Gaussian, 100, f32VecDim)
+			req.Requests[0].Vector.Id = ""
```
Comment (vankichi) [inline pkg/agent/core/ngt/handler/grpc/handler_test.go:4469]: How about refactor `genF32Req` for set no uuid with-param.
e.g. `genF32Req(vec, num, dim, noid)`
when sid is true, set uuid as "".

File: pkg/agent/core/ngt/handler/grpc/handler_test.go [hunk 5]
```diff
@@ -4302,6 +4319,702 @@ func Test_server_MultiInsert(t *testing.T) {
...
+				req.Requests[i].Vector.Id = ""
+			}
+
+			uuids := make([]string, 0, len(req.Requests))
+			for _, r := range req.Requests {
+				uuids = append(uuids, r.Vector.Id)
+			}
+
+			return test{
+				name: "Boundary Value Testing case 4.3: Fail to MultiInsert with 50 request with empty UUID (vector type is uint8)",
```
Comment (vankichi) [inline pkg/agent/core/ngt/handler/grpc/handler_test.go:4522]: I forgot why a half vectors with no uuid.
let me know the reason.

File: pkg/agent/core/ngt/handler/grpc/handler_test.go (hunk 4 above)
Comment (kevindiu) [reply pkg/agent/core/ngt/handler/grpc/handler_test.go:4469]: Becasue `genF32Req` will generate multiple insert requests, it is hard to decide which one needs empty ID and which one needs ID generation, I prefer the current style.

File: pkg/agent/core/ngt/handler/grpc/handler_test.go (hunk 3 above)
Comment (kevindiu) [reply pkg/agent/core/ngt/handler/grpc/handler_test.go]: thank you for your comment.
Since this value is copied from insert test, I also fixed the insert test

File: pkg/agent/core/ngt/handler/grpc/handler_test.go (hunk 5 above)
Comment (kevindiu) [reply pkg/agent/core/ngt/handler/grpc/handler_test.go:4522]: The reason I said was to ensure if part of the request will not be proceeded, and ensure if multiple errors occurred in one multiinsert request, how Vald handles it and no fatal error will be occurred

**Review by kmrmt — APPROVED** [review]

File: pkg/agent/core/ngt/handler/grpc/handler_test.go [hunk 6]
```diff
@@ -3742,6 +3743,22 @@ func Test_server_MultiInsert(t *testing.T) {
 		return req
 	}
 
+	genNInsertReq := func(num int, vec []float32) *payload.Insert_MultiRequest {
```
Comment (vankichi) [inline pkg/agent/core/ngt/handler/grpc/handler_test.go:3746]: It should be
```suggestion
	genMInsertReq := func(num int, vec []float32) *payload.Insert_MultiRequest {
```
or
```suggestion
	genMultiInsertReq := func(num int, vec []float32) *payload.Insert_MultiRequest {
```

File: pkg/agent/core/ngt/handler/grpc/handler_test.go (hunk 4 above)
Comment (vankichi) [reply pkg/agent/core/ngt/handler/grpc/handler_test.go:4469]: the kinds of `create request` function increases, so I thinks it would be better rename about create function.
e.g.
`genF32Req` -> `genF32InsertReq`
or
combine `genF32Req` and `genIntReq` to the `genInsertReq` with type option.

File: pkg/agent/core/ngt/handler/grpc/handler_test.go (hunk 6 above)
Comment (vankichi) [reply pkg/agent/core/ngt/handler/grpc/handler_test.go:3746]: I suppose it would be better
```go
	genMultiInsertReq := func(num int, type string) *payload.Insert_MultiRequest {
             // create vector
             // create request
        }
```

Comment (vankichi) [reply pkg/agent/core/ngt/handler/grpc/handler_test.go:3746]: duplicated implementation?
https://github.com/vdaas/vald/blob/fb5df3c3e355b9158ac2485b1701d7ef2ec5cabf/pkg/agent/core/ngt/handler/grpc/handler_test.go#L3710-L3726

File: pkg/agent/core/ngt/handler/grpc/handler_test.go (hunk 4 above)
Comment (kevindiu) [reply pkg/agent/core/ngt/handler/grpc/handler_test.go:4469]: The reason of separating genF32Req and genIntReq is because of the logic of generation is a little bit different (int request needs to cast to float32 to fit in our api). so I will rename it to genXXXInsertReq for now.

File: pkg/agent/core/ngt/handler/grpc/handler_test.go (hunk 6 above)
Comment (kevindiu) [reply pkg/agent/core/ngt/handler/grpc/handler_test.go]: do you mean it is duplicated with genIntReq? 
we can consider separate the create multiInsert request(line 3713-3725) to another function 🤔 
What do you think?

File: pkg/agent/core/ngt/handler/grpc/handler_test.go (hunk 4 above)
Comment (vankichi) [reply pkg/agent/core/ngt/handler/grpc/handler_test.go]: I could not understand.
You implemented a new multiInsertRequest generator function.
It seems similar.

> The reason of separating genF32Req and genIntReq is because of the logic of generation is a little bit different (int request needs to cast to float32 to fit in our api). so I will rename it to genXXXInsertReq for now.

I think we only have to implement using `swicth-case`. Isn't it correct?

Comment (kevindiu) [reply pkg/agent/core/ngt/handler/grpc/handler_test.go]: sorry I misread the code. yes you are right.I will refactor it now.

Comment (vankichi) [reply pkg/agent/core/ngt/handler/grpc/handler_test.go]: The difference between `genF32Req` and `genIntReq` is only a vector generator.

File: pkg/agent/core/ngt/handler/grpc/handler_test.go (hunk 6 above)
Comment (kevindiu) [reply pkg/agent/core/ngt/handler/grpc/handler_test.go]: I think this will be fixed by this refactoring 
https://github.com/vdaas/vald/pull/1645#discussion_r870980778

Comment (kevindiu) [reply pkg/agent/core/ngt/handler/grpc/handler_test.go]: it is not correct.
```
I suppose it would be better

	genMultiInsertReq := func(num int, type string) *payload.Insert_MultiRequest {
             // create vector
             // create request
        }
```
`genMultiInsertReq` generate multiInsert request with the same vector, so it should be OK with the current implement.

**Review by vankichi — APPROVED** [review]
